### ISSUE TITLE
Support for C#

### ISSUE THREAD

**Issue by mswilson4040** [looker-open-source/sdk-codegen#206]
This issue is a follow up from the discussion in https://github.com/looker-open-source/sdk-examples/issues/45

After talking with @jkaster, this issue is better suited in this repo.

The request is for Looker to provide a Swagger file that will work with C#. Currently, the only way to get C# working with the supplied swagger files (v3.0 - v4) comes with a significant amount of modifications to the swagger files. The process of the modification and what some of those modifications are can be seen in the linked issue (https://github.com/looker-open-source/sdk-examples/issues/45)

I would love to work with you guys on getting a C# supported swagger file out the door. If you have something that needs to be tested or any candidate builds, I would _love_ to be your Guinea Pig. We have a huge need for this!

**Comment by jkaster** [comment]
Hi Marc! You are **definitely** on our list to contact as we start working on the C# SDK. 

I'm currently working on a document describing how to [build your own SDK](https://github.com/looker-open-source/sdk-codegen/blob/jk/byosdk/byosdk.md) that will be used as a reference for both a C# SDK and our update to `LookR` (our R SDK). I plan to finish that document in the next two weeks, and kick off work on the R runtime then. It's possible we could also begin the C# runtime around that same time.

Question for you: What .NET runtime environments will you need for C#? 

My baseline plan was for .NET core since that's cross-platform and the largest target, but I figure most C# developers are still primarily Windows-based (as I was when I did C# development full-time). 

To help quantify and track status, I'll also be creating C# and R projects for this repository so we can get a high-level task list together for what's entailed in building a sustainable SDK.

**Comment by mswilson4040** [comment]
We have a range of projects that would consume this. Most of our C# code is running on .net core 3 or 3.1.

I think assuming baseline support for [.net standard 2.1](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) is pretty fair. 2.0 may be needed to support projects that are still running .net core 3.0, however.

Looking forward to any updates!

**Comment by jkaster** [comment]
Thanks for the answer and the link to .NET 3.0. I still keep up with my buddies from .NET for the most part, so I've been tracking what's going on there. It's definitely good to know your interests. That's really helpful.

**Comment by mswilson4040** [comment]
Hey, @jkaster just wanted to check in and see if there are any new updates on your current efforts. We have some upcoming projects that would strongly benefit from having a working looker client for c#

**Comment by jkaster** [comment]
Thanks for checking in, @mswilson4040! Your timing is impeccable: I was just going to give you an update today. I got bogged down in a major restructure of the codegen project. I still need to finish up the guide for building a new SDK (which will be a useful checklist for any new language). When that's done, we can start scheduling the milestones for the new SDKs.

Currently we're thinking work on C# and R, and possibly one more language, will start in Q3 2020. Because of your interest (and others), I'm trying to get C# started earlier than that but as is often the case, there's other prep work that needs to be done beforehand.

First step for me is finishing the BYOSDK doc, which will help our management scope the work for the new SDKs.

I appreciate your patience and really want to see our C# SDK become reality, too.

**Comment by mswilson4040** [comment]
Thanks for the update, @jkaster. Do you have any guess when we could get our hands on something? If Q3 2020 is a best-case scenario for the initial efforts, do you think by the end of the year you'll have something?

Happy to contribute if that would help get things moving quicker.

**Comment by jkaster** [comment]
I expect to be able to begin work on this in early Q3. I will **definitely** be asking for your engagement when the work starts.

First step is still finishing the BYO SDK doc, and it looks like I'll be able resume working on that June 1, 2020.

**Comment by terrysmith-fiworks** [comment]
Hi guys, I would also like to be kept up-to-date on this and could be a beta tester. We've used the swagger-generated API for nearly two years and have had constant problems with it including another one I'm working on now. We have very deep API integration with Looker (all server-to-server) including embedding. We're very ready for C# to be supported.

Regarding environments, we can't support .NET Core yet but are working towards it. We're currently on .NET Framework 4.8. An API that supports .NET Standard 2.0 should work for us.

**Comment by jkaster** [comment]
Hi @terrysmith-fiworks! Welcome to the party.

I'm making some forward progress on this again, and it's really great to get details on the required/desired .NET runtime environments.

@josephaxisa and myself are planning a 2.5 day development spike the week of June 15th, 2020 to make the C# SDK plans (and hopefully prototype progress) more concrete.

There will be a branch that week for that spike. 

I'm almost through the first draft of the aforementioned BYO SDK doc we'll be using as a rough task list. I just created the [C# SDK project](https://github.com/looker-open-source/sdk-codegen/projects/6) tracker and put in some placeholder tasks.

If you would like to create issues describing your specific run-time needs, that would be helpful.

**Comment by jkaster** [comment]
@josephaxisa and I were able to make good progress on the **Look#** project (catchy name, eh?). We actually got beyond conservative expectations in the [`csharp-rtl` branch](https://github.com/looker-open-source/sdk-codegen/tree/csharp-rtl) and were able to implement both the run-time library and the code generator, and successfully make a couple SDK calls. This is just our first spike, but you're welcome to take a look.

**Look#** is only set up for .NET Core 3.x right now. We've got lots of TODOs in the source, and even more in our heads. 

Please let us know what you think if you get a chance to take a look. Better to know early on if you like what's there so far!

P.S. Look# is the project code name and may not be the the actual name of the SDK when we're done.

**Comment by mswilson4040** [comment]
Hi @jkaster looks like the **csharp-rtl** branch was merged into master. I am revisiting this today to try and test it. I pulled the latest from master and after going through the steps of getting yarn installed, everything seems fine. But, once I run `yarn sdk csharp` (or any sdk), I get what seems to be an `package.json` issue with any entry file not being defined:

```
marc@Marcs-MacBook-Pro sdk-codegen % yarn sdk charp
yarn run v1.17.3
$ ts-node -O '{ "module": "commonjs", "target": "es2019" }' packages/sdk-codegen-scripts/src/sdkGen.ts charp
Error: Cannot find module '/Users/marc/dev/tmp/sdk-codegen/node_modules/@looker/sdk-codegen-utils/lib/index.js'. Please verify that the package.json has a valid "main" entry
    at tryPackage (internal/modules/cjs/loader.js:316:19)
    at Function.Module._findPath (internal/modules/cjs/loader.js:705:18)
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:969:27)
    at Function.Module._load (internal/modules/cjs/loader.js:864:27)
    at Module.require (internal/modules/cjs/loader.js:1044:19)
    at require (internal/modules/cjs/helpers.js:77:18)
    at Object.<anonymous> (/Users/marc/dev/tmp/sdk-codegen/packages/sdk-codegen-scripts/src/sdkGen.ts:30:1)
    at Module._compile (internal/modules/cjs/loader.js:1158:30)
    at Module.m._compile (/Users/marc/dev/tmp/sdk-codegen/node_modules/ts-node/src/index.ts:858:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1178:10)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by jkaster** [comment]
Sorry, I need to update steps in the readme for this experimental Look#. Do `yarn build` then try running `yarn sdk csharp` or `yarn sdk c#` again

**Comment by jkaster** [comment]
The long-term availability plan is to make Look# available via nuget

**Comment by jkaster** [comment]
Oh yeah! You shouldn't need to generate anything .. all the C# code is in `csharp`. But it really needs the nuget package. Hopefully I can sneak in some time for that soon. I've never pushed something to nuget before.

**Comment by mswilson4040** [comment]
Thanks, @jkaster! A nuget package would most certainly be the preferred method.

I was able to generate the C# code just fine. I pulled it into a separate project and might be confused about what the actual usage is. Do you have an example of doing something simple like logging in with clientId/clientSecret?

I think my main confusion is surrounding the below code. When I run it, `client.login` bombs out with

> System.ArgumentNullException: Path cannot be null. (Parameter 'path')
>    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
>    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
>    at System.IO.File.Open(String path, FileMode mode, FileAccess access, FileShare share)
>    at IniParser.FileIniDataParser.ReadFile(String filePath, Encoding fileEncoding)
>    at IniParser.FileIniDataParser.ReadFile(String filePath)
>    at Looker.RTL.ApiSettings.ReadConfig(String sectionName) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/ApiSettings.cs:line 72
>    at Looker.RTL.AuthSession._login(String sudoId) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 139
>    at Looker.RTL.AuthSession.Login(String sudoId) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 251
>    at Looker.RTL.AuthSession.GetToken() in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 238
>    at Looker.RTL.AuthSession.Authenticate(HttpRequestMessage request) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 100
>    at Looker.RTL.Transport.RawRequest(HttpMethod method, String path, Values queryParams, Object body, Authenticator authenticator, ITransportSettings

Here's my code

```
var apiSettings = new ApiSettings {BaseUrl = "<Base URL>"};
var transport = new Transport(apiSettings);
var authSession = new AuthSession(apiSettings, transport);
var client = new Looker31SDK(authSession);
var token = await client.login<AccessToken, Exception>("ClientId", "ClientSecret");
```

From the looks of it, it's expecting the looker.ini file to be present to read settings from rather than getting those from the ApiSettings object. Does that ini file need to exist on the consuming project or is that supposed to be exclusive to the sdk generation?

**Comment by jkaster** [comment]
For C#, I highly recommend you use `Looker40SDK` rather than 3.1 because 3.1 is guaranteed to have broken payloads from the API calls.

ApiSettings.ReadConfig() has a bug where it doesn't check to see if the file exists before attempting to read it.

You could create your own `ApiSettings()` descendant to override where to get the config settings, or just fix that "don't read a non-existent file!" issue. 

We'll fix that bug on line 72 of `ApiSettings.ReadConfig(String sectionName)` soon, flesh out the config options, and adding more tests next time we get a round for Look# but with a little bit of your own setup code and a fix for line you should be able to get this working.

**Comment by jkaster** [comment]
#253 has a fix for the non-existent `looker.ini` file so you may want that after it's been merged to master @mswilson4040

**Comment by mswilson4040** [comment]
@jkaster I pulled the latest and it looks like the _path can't be null_ is resolved. Not I get a new error:

```
System.Collections.Generic.KeyNotFoundException: The given key 'client_id' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Looker.RTL.AuthSession._login(String sudoId) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 140
   at Looker.RTL.AuthSession.Login(String sudoId) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 251
   at Looker.RTL.AuthSession.GetToken() in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 238
   at Looker.RTL.AuthSession.Authenticate(HttpRequestMessage request) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/AuthSession.cs:line 100
   at Looker.RTL.Transport.RawRequest(HttpMethod method, String path, Values queryParams, Object body, Authenticator authenticator, ITransportSettings options) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/Transport.cs:line 245
   at Looker.RTL.Transport.Request[TSuccess,TError](HttpMethod method, String path, Values queryParams, Object body, Authenticator authenticator, ITransportSettings options) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/Transport.cs:line 326
   at Looker.RTL.ApiMethods.AuthRequest[TSuccess,TError](HttpMethod method, String path, Values queryParams, Object body, ITransportSettings settings) in /Users/marc/dev/tmp/sdk-codegen/csharp/rtl/ApiMethods.cs:line 96
   at Looker.SDK.API40.Looker40SDK.login(String client_id, String client_secret, ITransportSettings options) in /Users/marc/dev/tmp/sdk-codegen/csharp/sdk/4.0/methods.cs:line 3622
   at ConsoleApp1.Program.Main(String[] args) in /Users/marc/dev/ConsoleApp1/ConsoleApp1/Program.cs:line 19
```

I may be confused on the **looker.ini** file placement/existence. It sounds like that needs to be present wherever you are consuming Look#. I placed the same **looker.ini** file that is in the sdk-codegen project at the consuming projects root, and still get "The given key 'client_id' was not present in the dictionary." I have the same basic code shown above, but pasting it again since it's been updated to use 4.0 instead of 3.1

```
var apiSettings = new ApiSettings {BaseUrl = "https://<domain>.looker.com:19999"};
var transport = new Transport(apiSettings);
var authSession = new AuthSession(apiSettings, transport);
 var client = new Looker40SDK(authSession);
try
{
    var token = await client.login("<client_id>", "<client_secret>"); // The given key 'client_id' was not present in the dictionary
}
catch (Exception ex)
{
    Console.WriteLine(ex.Message);
}
```

Is there a specific location that the **looker.ini** folder needs to be in the consuming project (if at all)? Not sure if you have a readily-available usage example. That might answer my questions.

**Comment by jkaster** [comment]
hey @mswilson4040 sorry you're still having trouble. For your example, you don't need to explicitly call `login()`. That happens automatically for you. try this:

```cs
try
{
    var user = await sdk.Ok(sdk.me());
}
```

instead in place of `var token ...`

**Comment by jkaster** [comment]
like this test: https://github.com/looker-open-source/sdk-codegen/blob/5eef1d2376366a04afe76fc5471976ce479f3005/csharp/rtl.Tests/SdkMethodsTests.cs#L44-L50

**Comment by jkaster** [comment]
Oh yeah, I think the run-time will try to read the `looker.ini` from your current directory? If not, it's probably from the exe directory. However, it looks like you are *not* using a `looker.ini` at all. In that case, you're going to need to create an ApiSettings.readConfig() override that returns your `client_id` and `client_secret` from some other source.
